### PR TITLE
Training chunk V2

### PR DESCRIPTION
**Pull request by Error323**
Please review carefully, @killerducky @glinscott 

* This was verified with the new tensorflow pipeline's ChunkParser V2.
* Reduced the number of games per chunk to 5 instead of 200
* Improve readability of variables

### CODE REVIEW

**Comment by killerducky** [comment]
I just quickly looked at the diff, this will make it so clients only write out the training file and upload to server once every 5 games? I think we want the clients to always upload training after every game. I often just ctrl-C the client, we will lose more data if only upload once every 5 games. 

Also it might help later to be able to analyze individual training files and know they are single games. For example I believe gcp analyzes the training data to see if they are anomalous (presumably due to bad OpenCL drivers).

**Comment by killerducky** [comment]
Oh I looked again and it seems dump_training with out_filename constructs an OutputChunker and calls the other dump_training function. When both those functions return the OutputChunker is destructed. The destructor calls flush_chunks(). So for the self-play case I don't think the m_game_count is going to do anything?

Am I following this right? Maybe there is some other code that also calls Training/OutputChunker?

**Comment by Error323** [comment]
Yes you are right. The client adds arguments to the train cmd. These ensure every chunk contains exactly 1 game.

I made the wrong assumptions earlier as I didn't know the train cmd could have arguments.

**Comment by glinscott** [comment]
The supervised path uses the number of games per chunk parameter.  Changing it to 5 is fine, it has no effect on the reinforcement learning games.

Overall, looks good!  Maybe one things to try is dumping both V1 and V2 and doing a check the same data is there?

File: src/Training.cpp [hunk 1]
```diff
@@ -158,6 +158,69 @@ void Training::dump_training(int game_score, const std::string& out_filename) {
...
+     * uint8 move_count (1 byte)
+     * int8 result (1 byte)
+     * self.v2_struct = struct.Struct('4s7696s896sBBBBBBBb')
+     */
+    static int VERSION = 1;
+
+    std::stringstream out;
+    for (const auto& step : m_data) {
+        // Store the binary version number (4 bytes)
+        out.write(reinterpret_cast<char*>(&VERSION), sizeof(VERSION));
```
Comment (evalon32) [inline src/Training.cpp]: 1. Portability: is it safe to assume identical byte order between client and server? (I guess it is if only x86-64 is relevant.)
2. The use of stringstream: would it work if you instead simply defined a `struct { int32 version; float probabilities[1924]; ... }`, copied all the data into it, and wrote it out as a blob? (Happy to send a followup cleanup PR if you prefer.)

Comment (Error323) [reply src/Training.cpp]: 1. Good point. Client side should always store data in little endian format. I know howto do this properly in Linux, but not Windows, any ideas?
2. I tried to keep the code as close to V1 as possible, hence the stringstream. One can view `dump_training` and `dump_training_v2` side by side and easily examine the differences.

Comment (evalon32) [reply src/Training.cpp]: 1. Don't know much about Windows programming myself, but I think Windows is always little endian?
2. Understood.

**Comment by Error323** [comment]
Bah unittesting the Chunker is a nightmare :(

**Comment by Error323** [comment]
Ok I'm experiencing data inconsistencies and I can't figure them out yet. I generated [training.0-v1.gz](https://github.com/glinscott/leela-chess/files/1820839/training.0-v1.gz) [training.0-v2.gz](https://github.com/glinscott/leela-chess/files/1820840/training.0-v2.gz) using the following settings:

```bash
./lczero --weights=1aa02cb1 --seed=1 --start='train 1 1' -t1 -q
```

For some reason the bitplanes 4,5,10 and 11 are incorrect for the first ply. I really don't understand why yet. Anyway, I was generating some games with it on the server, which convoluted the data now. Training did not go well as a result I think. Elo stayed similar to gen8. I will restart training and exclude binary data.

**Comment by evalon32** [comment]
I think I understand why. If you're comparing the ASCII output against a hex dump of the binary output, the bits in each byte are going to be reversed. The bits in other planes just happen to be symmetrical. That's because the ASCII version treats bit[0] as the most significant, but bit `bitset::to_ullong` treats it as the least significant. The order of bytes will match because Little Endian reverses it, but the order of bits in each byte is not.

**Comment by Error323** [comment]
This illustrates the problem quite clearly:

```c++
#include <iostream>
#include <sstream>
#include <bitset>
#include <cassert>

using bitboard = std::bitset<8*8>;

std::string hexdump(const bitboard plane) {
  std::stringstream out;
  for (auto bit = size_t{0}; bit + 3 < plane.size(); bit += 4) {
      auto hexbyte =  plane[bit]     << 3
                    | plane[bit + 1] << 2
                    | plane[bit + 2] << 1
                    | plane[bit + 3] << 0;
      out << std::hex << hexbyte;
  }
  out << std::dec;
  return out.str();
}

int main(void) {
  // 1011110110001110011110010110110000111010010110000100010000100001
  bitboard plane = 0xBD8E796C3A584421;
  std::cout << plane.to_string() << std::endl;

  auto s = hexdump(plane);
  std::cout << s << std::endl;
  printf("%lx\n", plane.to_ulong());
  
  std::stringstream ss;
  std::uint64_t b;
  ss << std::hex << s;
  ss >> b;
  printf("%lx\n", b);

  assert(b == plane.to_ulong());
  return 0;
}
```

Output on my machine:
```bash
$ g++ -std=c++14 main.cpp -O0 -g
[fhuizing@titan ~/Workspace/chess/leela-chess git:trainingchunksv2]
$ ./a.out 
1011110110001110011110010110110000111010010110000100010000100001
84221a5c369e71bd
bd8e796c3a584421
84221a5c369e71bd
a.out: main.cpp:36: int main(): Assertion `b == plane.to_ulong()' failed.
Aborted (core dumped)
```

@gcp @ywrt you may also want to take note here. Any reason why v1 was setup this way?

**Comment by Error323** [comment]
```c++
bitboard fix(bitboard plane) {
  for (auto i = 0; i < plane.size()/2; i++) {
    int tmp = plane[i];
    plane[i] = plane[63-i];
    plane[63-i] = tmp;
  }

  return plane;
}
```

Will fix it... I'm not sure how portable this is.

**Comment by mooskagh** [comment]
I haven't read what happening above the last comment, but it seems to me
that the function from the last comment can be written as:
std::reverse(plane.begin(), plane.end());

**Comment by Error323** [comment]
The bitboard is an std::bitset which doesn't have iterators.

**Comment by mooskagh** [comment]
Indeed, that's unexpected, as infamous vector<bool> does have iterators.
Sorry for the noise.

**Comment by Error323** [comment]
No worries :) I'm still quite puzzled by all this myself. The fix above doesn't fix the problem. This is now my script for testing when generating data with the command above. It reports that all bitplanes are wrong but the rest is good.

```python
#!/usr/bin/env python3

import sys
import chunkparser
import numpy as np
import gzip

with gzip.open(sys.argv[1], 'rb') as f:
    v1 = f.read()

with gzip.open(sys.argv[2], 'rb') as f:
    v2 = f.read()

parser = chunkparser.ChunkParser(chunkparser.ChunkDataSrc([v1, v2]))
gen1 = parser.convert_chunkdata_to_v2(v1)
gen2 = parser.convert_chunkdata_to_v2(v2)

for t1 in gen1:
    t2 = next(gen2)
    t1 = parser.convert_v2_to_tuple(t1)
    t2 = parser.convert_v2_to_tuple(t2)
    print(t1[0] == t2[0])
    print(t1[1] == t2[1])
    print(t1[2] == t2[2])

"""
pl1 = np.frombuffer(t1[0], dtype=np.uint8, count=120*8*8*1).reshape(1, 120, 8*8)
pl2 = np.frombuffer(t2[0], dtype=np.uint8, count=120*8*8*1).reshape(1, 120, 8*8)
for i in range(120):
    if (pl1[0][i] != pl2[0][i]).any():
        print(i)
        print(pl1[0][i])
        print(pl2[0][i])
"""

# drain the parser
for _ in parser.parse():
    pass
```

**Comment by Error323** [comment]
So in short: The bit order of the std::bitset needs to be reversed and then converted to big endian. This is a horrible fix :/ Linux only I'm afraid. If you have any good ideas please let me know!

**Comment by Error323** [comment]
Okay, I've verified full correctness for Linux using the scripts and configs above. We now always convert to little endian before writing to disk for *nix based systems. It would be great if someone could generate a training chunk with the above command: `./lczero --weights=1aa02cb1 --seed=1 --start='train 1 1' -t1 -q` Then I can verify correctness.

File: src/Training.cpp [hunk 2]
```diff
@@ -158,6 +158,84 @@ void Training::dump_training(int game_score, const std::string& out_filename) {
...
+            out.write(reinterpret_cast<char*>(&v), sizeof(v));
+        }
+
+        // 112 bitplanes (896 bytes)
+        int kFeatureBase = Network::T_HISTORY * 14;
+        assert(kFeatureBase*sizeof(step.planes.bit[0].to_ullong()) == 896);
+        for (int p = 0; p < kFeatureBase; p++) {
+            const auto& plane = fix_v2(step.planes.bit[p]);
+            auto val = htole64(plane.to_ullong());
+            assert(plane.size() % 4 == 0);
```
Comment (evalon32) [inline src/Training.cpp]: Methinks this code depends more specifically on `plane.size() == 64` (or `plane.size() == std::numeric_limits<unsigned long long>::digits`, if you want to be pedantic), because of the way it uses `to_ullong()`.

Comment (Error323) [reply src/Training.cpp]: Indeed, but the assert at line 209 implies this already. It's kind of unnecessary here I think?

Comment (evalon32) [reply src/Training.cpp]: I don't think it does. The assert at line 209 is equivalent to `assert(kFeatureBase*sizeof(unsigned long long) == 896);` It doesn't imply anything about `plane.size()`.

**Comment by evalon32** [comment]
@ywrt Per C++ standard, "When converting between an object of class **bitset\<N>** and a value of some integral type, bit position **pos** corresponds to the _bit value_ **1<<pos**." This is independent of architecture. Now, if you store the result of `to_ullong()` in memory, the _byte_ order is architecture-dependent. But the _bit_ order in each byte is not.

I think another way to look at this is: a bitplane doesn't inherently have most-significant or least-significant bits; those only come into play when combining bits into numbers.  On the one hand, `std::bitset::to_ullong()` treats bit 0 as least significant. On the other hand, `numpy.unpackbits()` treats bit 0 as most significant. So something has to give :)

**Comment by Error323** [comment]
```python
#!/usr/bin/env python3

import sys
import chunkparser
import numpy as np
import gzip

with gzip.open(sys.argv[1], 'rb') as f:
    v1 = f.read()

with gzip.open(sys.argv[2], 'rb') as f:
    v2 = f.read()

parser = chunkparser.ChunkParser(chunkparser.ChunkDataSrc([v1, v2]))
gen1 = parser.convert_chunkdata_to_v2(v1)
gen2 = parser.convert_chunkdata_to_v2(v2)

i = 0
s = 0
for t1 in gen1:
    t2 = next(gen2)
    t1 = parser.convert_v2_to_tuple(t1)
    t2 = parser.convert_v2_to_tuple(t2)
    p1 = np.frombuffer(t1[1], dtype=np.float32, count=1924)
    p2 = np.frombuffer(t2[1], dtype=np.float32, count=1924)
    assert(t1[0] == t2[0])
    assert(t1[2] == t2[2])
    a = np.argsort(p1[p1>0])
    b = np.argsort(p2[p2>0])
    assert((a == b).all())

# drain the parser
for _ in parser.parse():
    pass
```

Passes all tests for both Windows and Linux. I think we can squash & merge!
